src/middlewared/middlewared/plugins/cloud_sync.py

```diff
@@ -0,0 +1,522 @@
+
+        # Make sure only root can read it as there is sensitive data
+        os.chmod(self.tmp_file.name, 0o600)
+
+        config = dict(self.cloud_sync["credentials"]["attributes"], type=self.provider.rclone_type)
+        config = dict(config, **self.provider.get_credentials_extra(self.cloud_sync["credentials"]))
+        if "pass" in config:
+            config["pass"] = rclone_encrypt_password(config["pass"])
+
+        remote_path = None
+
+        if "attributes" in self.cloud_sync:
+            config.update(dict(self.cloud_sync["attributes"], **self.provider.get_task_extra(self.cloud_sync)))
+
+            remote_path = "remote:" + "/".join([self.cloud_sync["attributes"].get("bucket", ""),
+                                                self.cloud_sync["attributes"].get("folder", "")]).strip("/")
+
+            if self.cloud_sync.get("encryption"):
+                self.tmp_file.write("[encrypted]\n")
+                self.tmp_file.write("type = crypt\n")
+                self.tmp_file.write("remote = %s\n" % remote_path)
+                self.tmp_file.write("filename_encryption = %s\n" % ("standard" if self.cloud_sync["filename_encryption"]
+                                                                    else "off"))
+                self.tmp_file.write("password = %s\n" % rclone_encrypt_password(self.cloud_sync["encryption_password"]))
+                self.tmp_file.write("password2 = %s\n" % rclone_encrypt_password(self.cloud_sync["encryption_salt"]))
+
+                remote_path = "encrypted:/"
+
+        self.tmp_file.write("[remote]\n")
+        for k, v in config.items():
+            self.tmp_file.write(f"{k} = {v}\n")
+
+        self.tmp_file.flush()
+
+        return RcloneConfigTuple(self.tmp_file.name, remote_path)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.tmp_file:
+            self.tmp_file.close()
+
+
+async def rclone(job, cloud_sync):
+    # Use a temporary file to store rclone file
+    with RcloneConfig(cloud_sync) as config:
+        args = [
+            '/usr/local/bin/rclone',
+            '--config', config.config_path,
+            '-v',
+            '--stats', '1s',
+            cloud_sync['transfer_mode'].lower(),
+        ]
+
+        if cloud_sync['direction'] == 'PUSH':
+            args.extend([cloud_sync['path'], config.remote_path])
+        else:
+            args.extend([config.remote_path, cloud_sync['path']])
+
+        proc = await Popen(
+            args,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+        )
+        check_cloud_sync = asyncio.ensure_future(rclone_check_progress(job, proc))
+        await proc.wait()
+        if proc.returncode != 0:
+            await asyncio.wait_for(check_cloud_sync, None)
+            raise ValueError("rclone failed")
+        return True
+
+
+async def rclone_check_progress(job, proc):
+    RE_TRANSF = re.compile(r'Transferred:\s*?(.+)$', re.S)
```

Comment:
I know this was just copying code but what od you thinking of putting re.compile on module level, so we dont need to compile the regex on every call.